crates/typst-library/src/visualize/gradient.rs

```diff
@@ -1290,9 +1290,21 @@ fn sample_stops(stops: &[(Color, Ratio)], mixing_space: ColorSpace, t: f64) -> C
     if low == 0 {
         low = 1;
     }
+    if t == 1.0 {
+        while stops[low - 1].1 >= stops[low].1 {
```

Comment:
This seems like it could underflow. Overall, I'm not sure I fully understand the code. Performing an exact floating point comparison seems potentially problematic.

Unrelated to your PR, the binary search above looks like a standard binary search that could be replaced by the std version.